frontend/src/shared/validation/helpers/uri.ts

```diff
@@ -0,0 +1,59 @@
+/**
+  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License").
+  You may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+import { z, ZodString } from 'zod';
+import { errorUtil } from 'zod/lib/helpers/errorUtil';
+
+declare module 'zod' {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+    interface ZodString {
+        s3Uri(message?: errorUtil.ErrMessage): ZodString;
+        s3Prefix(message?: errorUtil.ErrMessage): ZodString;
+        s3Resource(message?: errorUtil.ErrMessage): ZodString;
+        datasetUri(message?: errorUtil.ErrMessage): ZodString;
+        datasetPrefix(message?: errorUtil.ErrMessage): ZodString;
+        datasetResource(message?: errorUtil.ErrMessage): ZodString;
+    }
+}
+
+ZodString.prototype.s3Uri = function (message?: errorUtil.ErrMessage): z.ZodString {
+    const defaultMessage = 'Must be a valid S3 URI';
+    return z.string().regex(/s3:\/\/(?<bucket>[a-z0-9][a-z0-9.-]{1,61}[a-z0-9])\/((?<prefix>([^/]+\/)+)?(?<object>[^/]+)?)?$/, message || defaultMessage);
+};
+
+ZodString.prototype.s3Prefix = function (message?: errorUtil.ErrMessage): z.ZodString {
+    const defaultMessage = 'Must be a valid S3 URI for a prefix that ends with `/`';
+    return z.string().regex(/s3:\/\/(?<bucket>[a-z0-9][a-z0-9.-]{1,61}[a-z0-9])\/(?<prefix>([^/]+\/)+)?$/, message || defaultMessage);
+};
+
+ZodString.prototype.s3Resource = function (message?: errorUtil.ErrMessage): z.ZodString {
+    const defaultMessage = 'Must be a valid S3 URI for an object (not ending with `/`)';
+    return z.string().regex(/s3:\/\/(?<bucket>[a-z0-9][a-z0-9.-]{1,61}[a-z0-9])\/(?<prefix>([^/]+\/)+)?(?<object>[^/]+)$/, message || defaultMessage);
+};
+
+ZodString.prototype.datasetUri = function (message?: errorUtil.ErrMessage): z.ZodString {
```

Comment:
Something to keep in mind with these is that we have documentation specifically enabling SAs to grant access to S3 buckets that are not datasets: https://github.com/awslabs/mlspace/blob/main/frontend/docs/admin-guide/manual-s3-permissions.md

Having these potentially constrains user from using custom S3 bucket access. We need to make sure we don't inadvertently prohibit usage of S3 buckets we tell Admins they can grant access to